srtcore/channel.cpp

```diff
@@ -90,86 +88,92 @@ m_iIpV6Only(-1)
-      addrinfo* res;
+void CChannel::open(const sockaddr_any& addr)
+{
+    createSocket(addr.family());
+    socklen_t namelen = addr.size();
 
-      memset(&hints, 0, sizeof(struct addrinfo));
+    if (::bind(m_iSocket, &addr.sa, namelen) == -1)
+        throw CUDTException(MJ_SETUP, MN_NORES, NET_ERROR);
 
-      hints.ai_flags = AI_PASSIVE;
-      hints.ai_family = m_iIPversion;
-      hints.ai_socktype = SOCK_DGRAM;
+    m_BindAddr = addr;
+    LOGC(mglog.Debug, log << "CHANNEL: Bound to local address: " << SockaddrToString(m_BindAddr));
 
-      if (0 != ::getaddrinfo(NULL, "0", &hints, &res))
-         throw CUDTException(MJ_SETUP, MN_NORES, NET_ERROR);
+    setUDPSockOpt();
+}
 
-      // On Windows ai_addrlen has type size_t (unsigned), while bind takes int.
-      if (0 != ::bind(m_iSocket, res->ai_addr, (socklen_t)res->ai_addrlen))
-      {
-          ::freeaddrinfo(res);
-          throw CUDTException(MJ_SETUP, MN_NORES, NET_ERROR);
-      }
-      memcpy(&m_BindAddr, res->ai_addr, res->ai_addrlen);
-      m_BindAddr.len = (socklen_t) res->ai_addrlen;
+void CChannel::open(int family)
+{
+    createSocket(family);
 
-      ::freeaddrinfo(res);
-   }
+    //sendto or WSASendTo will also automatically bind the socket
+    addrinfo hints;
+    addrinfo* res;
+
+    memset(&hints, 0, sizeof(struct addrinfo));
+
+    hints.ai_flags = AI_PASSIVE;
+    hints.ai_family = family;
+    hints.ai_socktype = SOCK_DGRAM;
+
+    int eai = ::getaddrinfo(NULL, "0", &hints, &res);
+    if (eai != 0)
+    {
+        // Controversial a little bit because this function occasionally
+        // doesn't use errno (here: NET_ERROR for portability), instead
+        // it returns 0 if succeeded or an error code. This error code
+        // is passed here then. A controversy is around the fact that
+        // the receiver of this error has completely no ability to know
+        // what this error code's domain is, and it definitely isn't
+        // the same as for errno.
+        throw CUDTException(MJ_SETUP, MN_NORES, eai);
+    }
+
+    // On Windows ai_addrlen has type size_t (unsigned), while bind takes int.
+    if (0 != ::bind(m_iSocket, res->ai_addr, (socklen_t)res->ai_addrlen))
+    {
+        ::freeaddrinfo(res);
+        throw CUDTException(MJ_SETUP, MN_NORES, NET_ERROR);
+    }
+    m_BindAddr = sockaddr_any(res->ai_addr, res->ai_addrlen);
+
+    ::freeaddrinfo(res);
 
-   HLOGC(mglog.Debug, log << "CHANNEL: Bound to local address: " << SockaddrToString(&m_BindAddr));
+    HLOGC(mglog.Debug, log << "CHANNEL: Bound to local address: " << SockaddrToString(m_BindAddr));
 
-   setUDPSockOpt();
+    setUDPSockOpt();
 }
 
-void CChannel::attach(UDPSOCKET udpsock)
+void CChannel::attach(int udpsock, const sockaddr_any& udpsocks_addr)
 {
-   m_iSocket = udpsock;
-   setUDPSockOpt();
+    // The getsockname() call is done before calling it and the
+    // result is placed into udpsocks_addr.
+    m_iSocket = udpsock;
+    m_BindAddr = udpsocks_addr;
```

Comment:
`UDPSOCKET` type name was descriptive. Why was it deleted?

Reply:
Then this is wrong. I started removal of this UDPSOCKET as it seemed to be an unnecessary alias to `int` - although it was actually created because of portability reasons. This will be restored.